### PR TITLE
AWS S3: Support for customer key server side encryption

### LINKED ISSUE #654: S3 Support Server Sside Encryption - Customer Keys
Currently, 
Server side encryption - customer keys (SSE-C) is not supported by S3 alpakka

Reference: http://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html

### PR DESCRIPTION
**Pull request by mlischetti**
fix #654 S3 support server side encryption - customer keys (SSE-C)

### CODE REVIEW

**Comment by ennru** [comment]
Could you add the corresponding methods to the Java API as well, please.

**Comment by ennru** [comment]
We merged a rather large PR for S3, could you re-base this PR, please.

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala [hunk 1]
```diff
@@ -63,9 +63,12 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
   // def because tokens can expire
   def signingKey = SigningKey(settings.credentialsProvider, CredentialScope(LocalDate.now(), settings.s3Region, "s3"))
 
-  def download(s3Location: S3Location, range: Option[ByteRange] = None): Source[ByteString, Future[ObjectMetadata]] = {
+  def download(s3Location: S3Location,
+               range: Option[ByteRange] = None,
+               sse: Option[ServerSideEncryption] = None): Source[ByteString, Future[ObjectMetadata]] = {
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: As this is internal API I'd prefer to not use default parameters (applies to several methods).

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Headers.scala [hunk 2]
```diff
@@ -71,30 +72,67 @@ object StorageClass {
...
-      RawHeader("x-amz-server-side-encryption", "aws:kms") ::
-      RawHeader("x-amz-server-side-encryption-aws-kms-key-id", kmsKeyId.get) ::
-      RawHeader("x-amz-server-side-encryption-context", context.get) ::
-      Nil
-    case _ => throw new IllegalArgumentException("Unsupported encryption algorithm.")
-  }
+sealed abstract class ServerSideEncryption {
+  def headers: immutable.Seq[HttpHeader]
+
+  def headers(request: S3Request): immutable.Seq[HttpHeader]
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Headers.scala]: It is not obvious what this `headers` method is supposed to do. Is  `applicableFor` a better name?

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala [hunk 3]
```diff
@@ -63,9 +63,12 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
...
     import mat.executionContext
-    val future = request(s3Location, rangeOption = range)
+    val s3Headers = S3Headers(sse.map(_.headers(GetObject)).getOrElse(Seq.empty))
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: a `fold` might be simpler

Comment (johanandren) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: or a `match` which is even more readable IMO (and does not allocate any extra lambdas)

Comment (mlischetti) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: LGTM I will changes _map + getOrElse_ by _fold_

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala (hunk 1 above)
Comment (mlischetti) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: @ennru Are you proposing simply remove the default parameters or split this method in 4 methods?

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Headers.scala (hunk 2 above)
Comment (mlischetti) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Headers.scala]: Before created this issue, **_ServerSideEncryption_** class had this method:
	_**def headers: immutable.Seq[HttpHeader]**_
and this method return all headers according with the specified server side encryption
but actually this headers are definied by the current request.
Maybe I should change the other method:
	**_def headers(request: S3Request):  immutable.Seq[HttpHeader]_**
by 
	**_def headersByRequest(request: S3Request):  immutable.Seq[HttpHeader]_**

Comment (ennru) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Headers.scala]: Yes, or `headersFor`?

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala (hunk 1 above)
Comment (ennru) [reply s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: All parameters (without defaults) in a single method to keep the internal API smaller.

**Pull request merged**